File: python/subproblem_setups/sp5.py
from linearSubproblemSltns import sp5_lib as sp5

import time as time

#Helper variables
p0 = sp5.np.array([1., 2., 3.])
p1 = sp5.np.array([1., 2., 3.])
p2 = sp5.np.array([1., 2., 3.])
p3 = sp5.np.array([1., 2., 3.])
k1 = sp5.np.array([1., 2., 3.])
k2 = sp5.np.array([1., 2., 3.])
k3 = sp5.np.array([1., 2., 3.])

#Begin solving
numTests = input("Enter the number of tests: ")

#Input formatting
numTests.strip()
numTests = int(numTests) #Convert from str to int

timeSum = 0  #Keep track of time for all operations
errorSum = 0 #Keep track of errorss
time1 = 0    #Start time
time2 = 0    #End time

#Run procedure
for i in range(numTests):
   #Perform setup
   sp5.sp5_setup(p0, p1, p2, p3, k1, k2, k3)
   #print("p1={}\r\np2={}\r\np3={}\r\nk1={}\r\nk2={}\r\nk3={}\r\n\np0={}\r\n".format(p1, p2, p3, k1, k2, k3, p0))
   time1 = time.perf_counter_ns()

   #Perform calculations
   theta1, theta2, theta3 = sp5.sp5_run(p0, p1, p2, p3, k1, k2, k3)
   time2 = time.perf_counter_ns()
   timeSum += time2-time1 #Collect time data
   errorSum += sp5.sp5_error(p0, p1, p2, p3, k1, k2, k3, theta1, theta2, theta3)

#Print results
print("\r\nResults: (Number of Tests: {})\r\n".format(numTests) + "-"*50)
print("Avg Time: {} ns, Total Time {} ns".format(timeSum/numTests, timeSum))
print("Average error margin: {}".format(errorSum/numTests))
print("Time in s: {}\r\n".format(timeSum/1000000000))

File: matlab/+subproblem_setups/sp_5.m
classdef sp_5
    methods (Static)
        function [P, S] = setup()
            P.p1 = rand_vec;
            P.p2 = rand_vec;
            P.p3 = rand_vec;
            
            P.k1 = rand_normal_vec;
            P.k2 = rand_normal_vec;
            P.k3 = rand_normal_vec;
            
            S.theta1 = rand_angle;
            S.theta2 = rand_angle;
            S.theta3 = rand_angle;
            
            P.p0 = -(rot(P.k1,S.theta1)*P.p1 ...
                   -rot(P.k2,S.theta2)*(P.p2+rot(P.k3,S.theta3)*P.p3));
        end

        function S = run(P)
            [S.theta1, S.theta2, S.theta3] = ...
                subproblem.sp_5(P.p0,P.p1,P.p2,P.p3,P.k1,P.k2,P.k3);
        end

        function e = error(P, S)
            e = NaN;
            for i = 1:length(S.theta1)
                e_i = norm(P.p0 + rot(P.k1,S.theta1(i))*P.p1 - rot(P.k2,S.theta2(i))*(P.p2+rot(P.k3,S.theta3(i))*P.p3));
                if i == 1
                    e = e_i;
                else
                    e = e + e_i;
                end
            end
        end
    end
end